docs: Tracking in a nutshell (#1399)
acts-project/acts: docs/tracking.md

uncertainties and correlations need to be taken into account as well. They
can be expressed as a $5\times 5$ covariance matrix like

\begin{equation*}
  C =
  \begin{bmatrix}
   \sigma^2(l_0)& \text{cov}(l_0,l_1) & \text{cov}(l_0, \phi) & \text{cov}(l_0, \theta) & \text{cov}(l_0, q/p) \\
   . & \sigma^2(l_1) & \text{cov}(l_1, \phi) & \text{cov}(l_1, \theta) & \text{cov}(l_1, q/p) \\
   . & . &  \sigma^2(\phi) & \text{cov}(\phi,\theta) & \text{cov}(\phi, q/p) \\
   . & . & . & \sigma^2(\theta) & \text{cov}(\theta, q/p) \\
   . & . & . & . & \sigma^2(q/p)
  \end{bmatrix}
\end{equation*}

Here, $\text{cov}(X,Y)$ is the covariance of variables $X$ and $Y$, while
$\sigma^2(X)$ are the regular variances. As the covariance matrix $C$ is
symmetric, only the upper right half is shown in the matrix above. The
uncertainties associated with the local position, as well as the momentum
direction are indicated in {numref}`parameters` (a) as an ellipse and a cone
around the momentum vector $\vec p$, respectively. 

(particle_propagation)=
## Particle propagation

:::{tip}
A dedicated description of the ACTS implementation of particle propagation can be found [here](propagation_impl).
:::

> A central for track reconstruction is the ability to calculate the trajectory

Author: ```suggestion
A central part of track reconstruction is the ability to calculate the trajectory
```

sketch of the way multiple Coulomb scattering affects the direction of a
particle trajectory. In addition, a shift in the transverse plane relative to
the incident direction can occur. As the scattering events occur in
statistically independent directions, the means of both the deflection and
offset tends toward zero as the number of scatters increases. Therefore, in the
numerical particle propagation, this can be accounted for by simply increasing
the uncertainties associated with the direction, depending on the amount of
material encountered.

On the other hand, there are interactions during which the particle loses
some of its energy. Relevant processes here are ionization, as well as
bremsstrahlung for light particles like electrons. For hadronic particles,
hadronic interactions with the nuclei of surrounding material is another
process of interest. In such hadronic interactions, the incoming particle
often disintegrates, and does not propagate further. Since the size of
ionization losses only fluctuates to a small degree for thin layers of
material, they can usually be accounted for by reducing the trajectory energy

> correspondingly. For bremsstrahlung, where fluctuations are much larger,

Reviewer: (and the effect can in addition not being modelled with a Gaussian distribution adequately)

significant performance bottleneck.

(layer_barrel)=
:::{figure} /figures/tracking/layer_barrel.svg
:width: 300px
:align: center
Sketch of the way sensitive elements are grouped into layers.
Shown is an $xy$-view of a number of sensors, arranged as in e.g. the ATLAS
silicon detector barrels. The grouping is based on their mounting radius.
The layers are indicated in different colors.
:::

As a compromise between modelling accuracy and performance, ACTS
uses a simplified geometry model. It
focusses on the sensitive elements, which are strictly needed, while passive
elements are discarded from the explicit description and approximated.
{numref}`geometry_detail` (b) shows such a simplified geometry. Here, the
sensitive elements are still shown in green, and other elements are greyed
out, indicating that they are discarded. The sensitive elements are then
grouped into layers, as sketched in {numref}`layer_barrel`. How exactly the
grouping occurs depends on the concrete experiment geometry. In some cases, the layers have the shape
of cylinder surfaces with increasing radii. This example is shown in the
figure in the transverse plane at radii $r_{1,2,3}$. In the endcaps, where

> modules are arranged on disks, these are used as the layer shape. An

Reviewer: are often arranged on disks

navigation between volumes. When the propagation has finished processing a
set of layers, it attempts to target the boundary surfaces. Once a boundary
surface is reached, the active volume is switched, and the next set of layers
is processed.

Care has to be taken to correctly model the passive material, that is
initially discarded with non-sensitive elements. For the material effects to
be correctly taken into account during particle propagation, the material is
projected onto dedicated material surfaces. These material surfaces are
spread across the detector geometry. Each layer is created with two
*approach surfaces* on either side. Their distance can be interpreted as
the thickness of the layer in question. Examples of these approach surfaces
can be found in {numref}`geometry_detail`, at the inner and outer radius.
Approach surfaces, and the boundary surfaces between volumes mentioned before,
are candidates to receive a projection of the surrounding material.
Additional artificial material layers can also be inserted to receive
projected material.

The projection procedure works by extrapolating test particles using the

> fully detailed simulation geometry. During the extrapolation, the material

Reviewer: Can we link the the material mapping documentation into this?

retrieved, and the corresponding modifications to the trajectory are
executed. In case material is supposed to be integrated in a continuous way
(as mentioned in [](#particle-propagation)), volumes can also store an
effective volumetric material composition, which is queried by the numerical
integration when needed. As the actual physical location of the detection
hardware can vary over time, possible misalignment of the sensors needs to be
handled correctly (see [](#alignment-of-particle-sensors)).

## Clusterization

The actual track reconstruction procedure itself starts with the conversion of
raw inputs that have been read out from the detector. In case of silicon
detectors, the readout can either be performed in a binary way, only recording
which segments fired, or the amount of charges measured in the segment can be
recorded, e.g. via *time-over-threshold* readout. In all cases, the readout is
attached to an identifier uniquely locating the segment on the corresponding
sensor.

As a next step, these raw readouts need to be *clustered*, in order to
extract an estimate of where particles intersect with the sensor. The general

> strategy of clustering algorithms follows the Connected Component Analysis (CCA)

Reviewer: Also here we could link the clustering / clustering documentation.

Author: I don't think we were concretely planning this, but I can create a stub. @gagnonlg would you be able to fill that out in a follow-up PR?

# Tracking in a nutshell

Track reconstruction is the process to recover the properties of a charged
particle from a set of measurements caused by interaction with some form of
sensitive detector. The goal is to find which measurements are likely to have
been caused by which particle, group them accordingly, and estimate the
associated trajectory. Such charged particle trajectories form the basic
input to the majority of higher-level reconstruction procedures in many
cases.

:::{figure} /figures/tracking/tracking.svg
:width: 400px
:align: center

> 
> Illustration of a track reconstruction chain starting from spacepoints to fully
> formed tracks

Reviewer 2: ```suggestion
Illustration of a track reconstruction chain starting from spacepoints to fully
formed tracks.
```

:::

This section provides a high-level view of a track reconstruction chain, and is largely based on [^phd_paul].

## Charged particle detection

The first step in the chain to reconstruct charged particles is their
detection using sensitive elements. Charged particle detection can be
achieved in a variety of ways with very different technologies. It can be
achieved by measuring the interaction of charged particles with matter. When
this occurs, the interacting particle typically ionizes the surrounding
material. Particle detectors make use of this fact by converting the
resulting charge into a measurable signal in various ways.

(segmentation)=
:::{figure} /figures/tracking/segmentation.svg
:align: center
:width: 400px

> Illustration of one-dimensional (a) and two-dimensional segmentation (b) of a silicon sensor.

Author: ```suggestion
Illustration of a one-dimensional (a) and a two-dimensional segmentation (b) of a silicon sensor.
```

can be included. Apart from the global reference frame, track quantities often
need to be represented with respect to a surface. This can be achieved with a
parametrization like

\begin{equation*}
  \vec x = \left(l_0, l_1, \phi, \theta, q/p, t\right)^T
\end{equation*}

although other parameter conventions exist as well.
{numref}`parameters` illustrates this choice of parameters. $l_0$, $l_1$ are
the local coordinates of the corresponding surface, $\phi \in [-\pi,\pi)$ and
$\theta \in [0,\pi]$ are the angles in the transverse and longitudinal
direction of the global frame, expressed with respect to the current location
along the trajectory, as indicated in {numref}`parameters` (b). $\theta$ is
the polar angle relative to the positive $z$-axis, and $\phi$ is the azimuth
angle in the transverse plane. Finally, $q/p$ combines the charge of the
particle with the inverse momentum. In {numref}`parameters` (a), the global
momentum vector $\vec p$ is shown, which can be recovered from the parameters
$\vec x$ using $\phi$, $\theta$ and $q/p$.

(parameters)=
:::{figure} /figures/tracking/parameters.svg
:width: 500px
:align: center

> 

Reviewer 2: ```suggestion
```

Illustration of the parametrization of a particle track with respect to a
two-dimensional surface. (a) shows the local position, global momentum and
their corresponding uncertainties. (b) displays the angles $\phi$ and $\theta$
in the transverse and longitudinal planes.

> 

Reviewer 2: ```suggestion
```

:::

(perigee)=
:::{figure} /figures/tracking/perigee.svg
:width: 200px
:align: center

> 

Reviewer 2: ```suggestion
```

changing fields, the corresponding differential equations of motions need to be
solved using numerical integration techniques.

### Numerical integration

In ACTS, numerical integration is done using the *Runge-Kutta-Nyström* (RKN) method.
Commonly used in the variant at fourth order, it describes how to calculate a
solution to an initial value problem that can be formulated generically like

$$
  \frac{dy}{dt} = f(t,y), \qquad y(t_0) = y_0,
$$

 where $y_0$ refers to the initial value of $y$ at $t_0$, and
$f(t,y)$ is the functional form describing the dynamics. The method then
successively approximates the analytical solution $y(t)$ in a stepwise
fashion. At each step $(t_n, y_n)$, the goal is effectively to approximate
the next step $y(t_{n+1})$. Using a step size $h$, the algorithm evaluates
the function $f$ at four points $k_{1-4}$:

$$
  \begin{aligned}
    k_1 &= f(t_n, y_n) \\
    k_2 &= f\left( t_n + \frac h 2, y_n + h \frac{k_1} 2 \right) \\
    k_3 &= f\left( t_n + \frac h 2, y_n + h \frac{k_2} 2 \right)\\
    k_4 &= f\left( t_n + h, y_n + hk_3 \right).
  \end{aligned}
$$

(rk)=
:::{figure} /figures/tracking/rk.svg
:width: 400px
:align: center

> 

Reviewer 2: ```suggestion
```

    \vec r_{n+1} &= \vec r_n + h \vec T_n + \frac{h^2}{6} (k_1 + k_2 + k_3).
  \end{aligned}
$$

A strategy exists to dynamically adapt the step size according to the magnetic
field strength, with the definition of a target accuracy that the algorithm
tries to achieve. Here, the step size $h$ will successively be decreased and
the approximation recalculated until the accuracy goal is achieved. Even with
these additional calculations, the approach is still preferable over a
consistently low step size.

### Covariance transport

Aside from the prediction of the track parameters at a given path length, a key
ingredient to many dependent applications are the uncertainties in the form of
the associated covariance matrix $C$. Conversions between covariance matrices
$C^i\to C^f$ can generally be achieved like

$$
  C^f = J \cdot C^i \cdot J^T,
$$

using the Jacobian matrix

$$
  J = \begin{bmatrix}
    \frac{\partial l_0^f}{\partial l_0^i} & \cdots & \frac{\partial l_0^f}{\partial (q/p)^i} \\
    \vdots & \ddots & \vdots \\
    \frac{\partial (q/p)^f}{\partial l_0^i} & \cdots & \frac{\partial (q/p)^f}{\partial (q/p)^i}
  \end{bmatrix},

> $$

Author: I see your point, but I think this would make the block here more complicated than it needs to be. I'll keep it like this for now.

between initial and final parameters $\vec x^i$ and $\vec x^f$. The
task therefore becomes calculating the necessary Jacobians to achieve correct
transformation.

One part is the transformation between different
coordinate systems, but at the same location along the trajectory. For this
purpose, generic Jacobians can be calculated between each coordinate system
type, and a common coordinate system. The common coordinate system used for
this purpose is the curvilinear frame, which consists of the global direction
angles, and a plane surface located at the track location, with its normal

> aligned with the track momentum. By using Jacobians to the curvilinear frame

Author: No I'm referring to the normal of the plane surface being aligned with the track momentum.

> angles, and a plane surface located at the track location, with its normal
> aligned with the track momentum. By using Jacobians to the curvilinear frame

Reviewer 2: Ah thank you, maybe we could make it explicit:

```suggestion
angles, and a plane surface located at the track location, with the normal of the plane surface
aligned with the track momentum. By using Jacobians to the curvilinear frame
```

and the corresponding inverse matrices, conversions between any two
coordinate systems can be performed.

The second part is the calculation of the evolution of the covariance matrix
during the propagation between surfaces. To this end, a semi-analytical
method which calculates the effective derivatives between two consecutive
RKN steps can be used. By accumulating the Jacobian
matrices calculated for each step, the effective Jacobian between the
starting point and the destination can be obtained.

## Material effects

Charged particles interact with matter as they pass through it. Since
particle detectors inevitably consist of some form or material, this effect
cannot be completely avoided. By building tracking detectors as light as
possible, and arranging passive components, such as services and support
structures carefully, the material a particle encounters before being
measured can be reduced. Charged particles traversing any form of matter
undergo elastic and inelastic interactions with the atomic structure of the
material, depending on the particle properties.

(multiple_scattering)=

:::{figure} /figures/tracking/multiple_scattering.svg
:width: 200px
:align: center

> 

Reviewer 2: ```suggestion
```

:::

A detailed model of the geometry of an experiment is required for tracking. In
many cases, external information is needed to associate a sensitive element
with a position and rotation in the laboratory frame. In case of silicon
sensors, the intrinsic information captured by the sensor is restricted to the
measurement plane. Using a transformation matrix, this local measurement can be
turned into a global one.

Full simulation using tools like Geant4 are frequently used in HEP.
It includes its own geometry
description framework. For the precise simulation of particle interactions
with the detector, this geometry modelling
is highly detailed. Even very small details of the
physical hardware can be crucial, and are often included in the geometry
description. An example for this are readout chips on silicon sensors, or cooling
elements. {numref}`geometry_detail` (a) shows a sketch of such a detailed
geometry description. Shown as an example is a *layer* of silicon
sensors in a barrel configuration. The green rectangles represent the actual
sensitive surfaces, while other elements include cooling, readout and other components.

(geometry_detail)=

> 

Reviewer 2: ```suggestion
```

:::{figure} /figures/tracking/geometry_detail.svg
:align: center

Sketch of the way a fully detailed simulation geometry (a) models passive
elements, in addition to the sensitive elements shown in green. (b) shows a
simplified version, where all non-sensitive elements are approximated.

> 

Reviewer 2: ```suggestion
```

:::

In the majority of cases in track reconstruction, this detailed
geometry is unnecessary. During track reconstruction, the aforementioned
associated information needs to be accessible for measurements, so all
sensitive elements need to be included in some form. Passive elements, on the
other hand, are only required to factor in material interaction effects (see
[](#particle-propagation)). Moreover, the fully detailed geometry comes
at the disadvantage of introducing significant overhead during navigation. In
this process, an algorithm attempts to figure out which elements the particle
propagation needs to target, as the trajectory is likely to intersect them.
With a geometry description this precise, the navigation process becomes a
significant performance bottleneck.

(layer_barrel)=
:::{figure} /figures/tracking/layer_barrel.svg
:width: 300px
:align: center

> 

Reviewer 2: ```suggestion
```

Sketch of the way sensitive elements are grouped into layers.
Shown is an $xy$-view of a number of sensors, arranged as in e.g. the ATLAS
silicon detector barrels. The grouping is based on their mounting radius.
The layers are indicated in different colors.

> 

Reviewer 2: ```suggestion
```

:::

As a compromise between modelling accuracy and performance, ACTS

> uses a simplified geometry model . It

Reviewer 2: ```suggestion
uses a simplified geometry model. It
```

focusses on the sensitive elements, which are strictly needed, while passive
elements are discarded from the explicit description and approximated.
{numref}`geometry_detail` (b) shows such a simplified geometry. Here, the
sensitive elements are still shown in green, and other elements are greyed
out, indicating that they are discarded. The sensitive elements are then
grouped into layers, as sketched in {numref}`layer_barrel`. How exactly the

> grouping occurs depends on the concrete experiment geometrym. In some cases, the layers have the shape

Reviewer 2: ```suggestion
grouping occurs depends on the concrete experiment geometry. In some cases, the layers have the shape
```

of cylinder surfaces with increasing radii. This example is shown in the
figure in the transverse plane at radii $r_{1,2,3}$. In the endcaps, where
modules are arranged on disks, these are used as the layer shape. An
illustration of endcap disk layers can be found in {numref}`layer_ec`, where
six disks are located at six distinct positions in $\pm z_{1,2,3}$, and
shown in different colors.

(layer_ec)=
:::{figure} /figures/tracking/layer_ec.svg
:width: 400px
:align: center

> 

Reviewer 2: ```suggestion
```

Sketch of the way sensitive elements are grouped into layers.
Shown is a view of a number of sensors, arranged as in e.g. the ATLAS
silicon detector endcaps. They are grouped into disks based on their
mounting position in $z$. The layers are indicated in different colors.

> 

Reviewer 2: ```suggestion
```

:::

During particle propagation, the navigation makes use of this layer
system. Each layer contains a binned structure, which maps a bin to a set
of sensitive surfaces that overlap with the bin area. This is illustrated in
{numref}`geo_binning`, where the left picture shows the sensitive surface
structure of an exemplary endcap disk. The picture on the right overlays the
binning structure that can be used to enable fast retrieval of compatible
sensitive surfaces. By performing a simple bin lookup, the navigation can
ascertain which sensors it needs to attempt propagation to.

(geo_binning)=
:::{figure} /figures/tracking/surface_array.svg
:width: 400px
:align: center

> 

Reviewer 2: ```suggestion
```

Illustration of the binning structure that is used to subdivide
layer surfaces. (a) shows two sensor rings of
different radii grouped into one disk layer. (b)
overlays the binning structure that the navigation queries for compatible
surfaces.

> 

Reviewer 2: ```suggestion
```

surface. Finally, the projection is averaged. The exact number and placement
of the material surfaces has to be optimized to yield a sufficiently accurate
representation of the inactive material in the detector.

The numerical integration uses these projected material surfaces. Whenever
such a surface is encountered in the propagation, the material properties are
retrieved, and the corresponding modifications to the trajectory are
executed. In case material is supposed to be integrated in a continuous way
(as mentioned in [](#particle-propagation)), volumes can also store an
effective volumetric material composition, which is queried by the numerical
integration when needed. As the actual physical location of the detection
hardware can vary over time, possible misalignment of the sensors needs to be
handled correctly (see [](#alignment-of-particle-sensors)).

## Clusterization

The actual track reconstruction procedure itself starts with the conversion of
raw inputs that have been read out from the detector. In case of silicon
detectors, the readout can either be performed in a binary way, only recording
which segments fired, or the amount of charges measured in the segment can be

> recorded, e.g. via  *time-over-threshold* readout. In all cases, the readout is

Reviewer 2: ```suggestion
recorded, e.g. via *time-over-threshold* readout. In all cases, the readout is
```

Filter](#combinatorial-kalman-filter) (CKF), uses the *Kalman formalism*.
Originally developed for monitoring and steering mechanical systems, it can
also be used to iteratively calculate a track estimate. After a set of track
candidates has been assembled and filtered (see [](#ambiguity-resolution)), an
additional track fit is usually performed to extract the best estimate of the
track. The Kalman formalism can also be used for this, with the addition of a
smoothing step that has certain benefits.  Other fit strategies exist, such as
a global $\chi^2$ fit that minimizes the distances between track-sensor
intersections and measurements on all sensors at the same time. One drawback of
this method is the necessity to invert very large matrices, which is
computationally expensive.

In a track fit, the Kalman formalism can be shown to yield optimal estimates
for Gaussian uncertainties. This assumption breaks down when effects like
bremsstrahlung come into play. An extension of the Kalman Filter (KF) exists
that relies on the individual propagation of a set of trajectories, instead of
a single one, to model these biased uncertainties by a sum of Gaussian

> components. This Gaussian Sum Filter (GSF) achieves better results when
> fitting particles such as electrons, likely to undergo bremsstrahlung, and is
> deployed in e.g. the ATLAS tracking chain.
> 
> :::{attention}
> TODO: Add GSF link
> :::

Reviewer 2: Did I understand correctly, what you inteded with the TODO?

```suggestion
components. This Gaussian Sum Filter [(GSF)](gsf)achieves better results when
fitting particles such as electrons, likely to undergo bremsstrahlung, and is
deployed in e.g. the ATLAS tracking chain.
```

Author: I actually wanted to keep this in to remind us. What do you think?

Reviewer 2: Sure let's keep it.

$$
  \vec m_k = \mathbf H_k \vec x_k + \epsilon_k.
$$

Here, $\mathbf H_k$ is the *measurement mapping function*, which
transforms the state vector into the measurement space. In the ideal case,
this purpose can be achieved by a simple projection matrix, which extracts a
subspace of the state vector. Additionally, $\epsilon_k$ represents the
measurement uncertainty.

The Kalman fit process is divided into different phases:

1. **Prediction** of the state vector at the next step $k+1$ based on the information at the current step $k$.
2. **Filtering** of the prediction by incorporating the measurement associated to the step
3. **Smoothing** of the state vector by walking back the steps and using information for the subsequent step $k+1$ to improve the estimate at the current step $k$.

An illustration of these concepts is found in {numref}`kalman_filter`. Here,
a series of three sensors is shown with measurements on them. The KF
then predicts the track parameters at an intersection, shown in blue.
Subsequently, a filtered set of parameters is calculated as a mixture between
the measurement and the prediction. Not shown in this picture is the
smoothing step.

(kalman_filter)=

> 

Reviewer 2: ```suggestion
```

:::{figure} /figures/tracking/kalman.svg
:width: 400px
:align: center

> 

Reviewer 2: ```suggestion
```

filtering are shown. The filtering updates the prediction with information from
the measurement.
:::

In many cases, the first two phases run in tandem, with prediction and
filtering happening alternatingly at each step. The smoothing phase is
launched once the last measurement has been encountered.
Starting from a state $k$, first, a prediction of the state vector at the
next measurement location is obtained via

(kf_pred)=
$$
  \vec x_k^{k-1} = \mathbf F_{k-1} \vec x_{k-1},
$$

with the linear transport function from above. $\vec x_k^{k-1}$ is
the prediction of the state vector at step $k$ based on step $k-1$. The next
stage is the filtering. Here, the state vector is refined by taking into
account the measurement at the current step. Following one of two variants of
filtering from [^Fruhwirth:1987fm], the gain matrix formalism, the state
vector is updated like

$$
  \vec x_k = \vec x_k^{k-1} + \mathbf K_k \left( \vec m_k - \mathbf H_k \vec x_k^{k-1} \right),
$$

with the *Kalman gain matrix*

$$
  \mathbf K_k = \mathbf C_k^{k-1} \mathbf H_k^\mathrm{T}
    \left(
      \mathbf V_k + \mathbf H_k \mathbf C_k^{k-1} \mathbf H_k^\mathrm{T}
    \right)^{-1}

>     %= \mathbf C_k \mathbf H_k^\mathrm{T} \mathbf G_k

Reviewer 2: Was it in your intention to comment this line out?
It also happens below in some places.

Author: I don't remember, this was for my thesis. Probably fine to remove the comments.
```suggestion
```

  ,
$$

can be defined, depending on which state estimate is compared with
the measurement $\vec m_k$. Using the filtered residual, an effective
$\chi^2$ increment

$$
  \chi^2_+ = \vec r_k^\mathrm{T} 
  % \mathbf R_k^{-1} 
  \left[ \left( \mathbb 1 - \mathbf H_k  \mathbf K_k \right)  \mathbf V_k \right]^{-1}
  \vec r_k
$$

can be calculated. The global $\chi^2$ of the track candidate can
be calculated as the sum of all $\chi^2_+$ across the steps. Measurements
with a large $\chi^2_+$ are considered as outliers, which have low
compatibility with the trajectory. By branching out for measurements below a
certain $\chi^2_+$, and following the branches, a tree-like structure of
compatible track candidates originating from a track seed is assembled. This
feature is shown in {numref}`tracking_ckf`, which displays a circular
trajectory, and a set of iteratively assembled track candidates. Basic
quality criteria can be applied at this stage, to remove bad candidates. A
dedicated [](#ambiguity-resolution).
selects the candidates most likely to belong to real particle tracks.

(tracking_ckf)=
:::{figure} /figures/tracking/finding.svg
:width: 300px
:align: center

> 

Reviewer 2: ```suggestion
```

There is currently no general-purpose ambiguity resolution implemented in ACST.
:::

Due to the combinatorial nature of track finding, and to achieve high
efficiencies, this set of candidates is often large, and contains a
non-negligible fraction of *fake* candidates. These fake candidates are either
completely combinatorial, or arise from real particle measurements with
combinatorial additions. Track candidates coming from a single seed necessarily
share a common stem of measurements. Measurements can potentially also be
shared between candidates from different seeds.

One possibility to resolve this (as is done in e.g. ATLAS tracking) is an
ambiguity resolution algorithm, that attempts to filter out as many undesirable
tracks as possible. This is implemented by means of a scoring function, that
combines properties of the track parameters. Higher scores are correlated with
a larger probability to be a desirable track candidate. A larger number of hits
results in an increase in the score, as longer compatible hit chains are less
likely to be random combinations. On the other hand, missing hits in sensors
where a hit was expected negatively impact the score.  Experiment specific

> scoring of hits from different subsystems is also implemented.  The overall

Reviewer 2: ```suggestion
scoring of hits from different subsystems is also implemented. The overall
```

## Vertex reconstruction

:::{attention}
A dedicated description of the vertexing as implemented in ACTS, and the 
various components that exist will be added shortly.
:::

One very immediate use case for tracks is the reconstruction of interaction
vertices in the event. These are needed for many aspects of higher-level
reconstruction algorithms. Vertices are used in different ways. As shown in
{numref}`vertexing_illust`, a single event can have a primary vertex, usually
associated with the hard-scatter event, as well as a number of secondary
vertices. Typically, the primary vertex will be located in the luminous
region. The reconstruction of secondary vertices can help with the
identification of particles, for instance in the form of *$b$-tagging*
jets originating from $b$-quarks. Here, a characteristic displaced vertex
located inside a jet is a sign for a $b$-hadron decay. Additionally, pile-up
interactions can produce vertices that are not associated with the
interaction of interest. Signatures from pile-up interactions can be
rejected, if they can be associated to a pile-up vertex.

(vertexing_illust)=
:::{figure} /figures/tracking/vertexing.svg
:width: 400px
:align: center

> 

Reviewer 2: ```suggestion
```

Illustration of a set of three vertices in a proton-proton
collision. A primary vertex and a secondary vertex are shown in addition to
a pile-up vertex.
:::

Vertex reconstruction can be divided into two stages: vertex finding and
vertex fitting, in close analogy to the reconstruction of tracks themselves.
Reconstruction begins with a seeding stage. Vertex seeding can be performed
in a number of ways. Algorithms differ for primary and secondary vertex
reconstruction. For primary vertex seeding, one option is an algorithm, which
uses a histogram approach to cluster tracks in the $z$-axis. This is based on
the assumption that primary vertices will align with the beam spot. Other
approaches include the use of a Gaussian track density to identify regions
with high track multiplicity as vertex seeds. For secondary vertexing,
seeds are formed from pairs of reconstructed tracks, as the constraint to the
beam spot does not apply.

With a set of vertex seeds determined, tracks compatible with the vertex are
selected. Using a linearized approximation of the helical trajectories in the

> vicinity of the vertex , a fitting procedure based on the Kalman formalism can

Reviewer 2: ```suggestion
vicinity of the vertex, a fitting procedure based on the Kalman formalism can
```
